File: term4/matlog/C/cpp/src/main.cpp
#include "main.h"


Var a = new variable("a"), b = new variable("b"), c = new variable("c");
const std::vector<Expr> axioms {
    new implication(a, new implication(b, a)),
        new implication(new implication(a, b), new implication(new implication(a, new implication(b, c)), new implication(a, c))),
        new implication(a, new implication(b, new conjunction(a, b))),
        new implication(new conjunction(a, b), a), new implication(new conjunction(a, b), b),
        new implication(a, new disjunction(a, b)), new implication(b, new disjunction(a, b)),
        new implication(new implication(a, c), new implication(new implication(b, c), new implication(new disjunction(a, b), c))),
        new implication(new implication(a, b), new implication(new implication(a, new negation(b)), new negation(a))),
        new implication(new negation(new negation(a)), a)
};


size_t isAxyp(const std::vector<Expr>& lft, const Expr expr, bool flag) {
    for (size_t i = 0; i < lft.size(); i++) {
        ntmCLear();
        if (lft[i]->equals(expr, flag)) return i + 1;
    }
    return { 0 };
}

size_t isAx(const Expr expr) {
    return isAxyp(axioms, expr, false);
}

size_t isHyp(const std::vector<Expr>& lft, Expr expr) {
    return isAxyp(lft, expr, true);
}

Expr reinit(std::vector<Expr>& lft, Expr e, bool insert = true) {
    //vAn.push_back({ lft, e });

    Impl ptr = dynamic_cast<Impl>(e);
    while (ptr != nullptr) {
        if (insert) {
            lft.push_back(ptr->left);
        }
        e = ptr->right;
        ptr = dynamic_cast<Impl>(e);
    }
    return e;

    //vnytriDeda.push_back({ lft, e });
}

bool cc(const std::vector<Expr>& sm, const std::vector<Expr>& bg) {
    std::vector<bool> dop(bg.size());
    if (sm.size() <= bg.size()) {
        std::fill(dop.begin(), dop.end(), false);
        for (Expr sh : sm) {
            size_t j{ 0 };
            for (; j <= bg.size(); j++) {
                if (!dop[j] && j != bg.size() && sh->equals(bg[j])) {
                    dop[j] = true;
                    break;
                }
            }
            if (j == bg.size() + 1) return false;
        }
        return true;
    }
    return false;
}

upair mp(const tvec& vAn, size_t i) {
    for (size_t j = i; j < std::string::npos; j--) {
        Impl dop = dynamic_cast<Impl>(vAn[j].second);
        if (dop && dop->right->equals(vAn[i].second) && cc(vAn[j].first, vAn[i].first)) {
            for (size_t k = i; k < std::string::npos; k--) {
                if (k != j && dop->left->equals(vAn[k].second)) {
                    if (cc(vAn[k].first, vAn[i].first)) {
                        return { j, k };
                    }
                }
            }
        }
    }
    return { -1, -1 };
}

void rec_degradation(const tvec& vAn, const std::vector<tpair>& globalAns, 
    std::vector<tpair>& gloablTmo, Expr e, size_t num) {
    for (size_t i = num + 1; i < globalAns.size(); i++) {
        if (globalAns[i].first->equals(e)) { 
            rec_degradation(vAn, globalAns, gloablTmo, globalAns[i].first, i);
            goto skip;
        }
    }
    for (Expr h : vAn.back().first) {
        if (e->equals(h)) {
            gloablTmo.push_back({ h, HYPH });
            goto skip;
        }
    }
    if (isAx(e) == 0) {
        for (size_t i = num + 1; i < globalAns.size(); i++) {
            Impl dop = dynamic_cast<Impl>(globalAns[i].first); 
            if (dop && dop->right->equals(e)) {
                for (size_t j = num + 1; j < globalAns.size(); j++) {
                    if (i != j && dop->left->equals(globalAns[j].first)) { 
                        gloablTmo.push_back({ e, MOP });
                        rec_degradation(vAn, globalAns, gloablTmo, globalAns[j].first, j);
                        rec_degradation(vAn, globalAns, gloablTmo, globalAns[i].first, i);
                        goto skip;
                    }
                }
            }
        }
    }
    gloablTmo.push_back({ e, AXIOM });
    skip:;
}

#define REPEAT(E, V) Impl vv = dynamic_cast<Impl> (E); \
                        if (!vv) break;                \
                        E = vv->right;

tppair dd(const tvec& vAn, size_t num) {
    std::vector<Expr> dop = vAn[num].first;
    Expr expr = reinit(dop, vAn[num].second);

    for (size_t i = num; i--;) {
        std::vector<Expr> vec(vAn[i].first);

        if (expr->equals(reinit(vec, vAn[i].second)) && cc(vec, dop)) {
            ttvec dop2;
            Expr tmp = vAn[i].second;
            for (;;) {
                expr = vAn[num].second;
                std::vector<Expr> kostil;
                for(;;) {
                    if (expr->equals(tmp)) {
                        for (size_t j = 0; j < kostil.size(); j++) {
                            dop2.push_back({ true, kostil[kostil.size() - j - 1] });
                        }
                        return { i + 1, dop2 };
                    }

                    REPEAT(expr, kostil)
                    kostil.push_back(vv->left);
                }

                REPEAT(tmp, dop2)
                dop2.emplace_back(false, vv->left);
            }
            break;
        }
    }
    return { 0, {} };
}

void GetFirstType(const std::pair<bool, Expr>& it, std::vector<tpair>& tmp) {
    Impl t1 = new implication(it.second, it.second);
    Impl t2 = new implication(it.second, t1);
    Impl t3 = new implication(it.second, new implication(t1, it.second));
    Impl t4 = new implication(t3, t1);

    {
        tmp.push_back({ t2, AXIOM });
        tmp.push_back({ new implication(t2, t4), AXIOM });
        tmp.push_back({ t4, MOP });
        tmp.push_back({ t3, AXIOM });
        tmp.push_back({ t1, MOP });
    }
}
void GetSecondType(const std::pair<bool, Expr>& it, std::vector<tpair>& tmp, Expr expr) {
    Impl t1 = new implication(it.second, expr);

    {
        tmp.push_back({ new implication(expr, t1), AXIOM });
        tmp.push_back({ expr, HYPH });
        tmp.push_back({ t1, MOP });
    }
}
void GetThirdType(const std::pair<bool, Expr>& it, std::vector<tpair>& tmp, Expr expr, 
    size_t num, const std::vector<tpair>& globalAns) {
    size_t cntr{ 0 }, dop{ 0 };
    for (size_t i = num + 1; i < globalAns.size() && !cntr; ++i) {
        Impl ab = dynamic_cast<Impl>(globalAns[i].first);
        if (ab && ab->right->equals(globalAns[num].first)) {
            for (size_t j = num + 1; j < globalAns.size(); ++j) {
                if (i != j && ab->left->equals(globalAns[j].first)) {
                    cntr = j + 1;
                    dop = i + 1;
                    break;
                }
            }
        }
    }
    Impl t1 = new implication(it.second, expr);
    Impl t2 = new implication(new implication(it.second, globalAns[dop - 1].first), t1);
    tmp.push_back({ new implication(new implication(it.second, globalAns[cntr - 1].first), t2), AXIOM });
    tmp.push_back({ t2, MOP });
    tmp.push_back({ t1, MOP });
}

void prove(const tvec& vAn, std::vector<tpair>& globalAns, size_t num) {
    Expr expr = vAn[num].second;
    for (Expr it : vAn[num].first) { 
        if (it->equals(expr)) { 
            globalAns.push_back({ expr, HYPH });
            return; 
        } 
    }
    if (isAx(expr) > 0) { 
        globalAns.push_back({ expr, AXIOM });
        return; 
    }
    upair mod = mp(vAn, num);
    if (mod.first != -1 && mod.second != -1) {
        globalAns.push_back({ expr, MOP });
        prove(vAn, globalAns, mod.first);
        prove(vAn, globalAns, mod.second);
    }
    else {
        auto dDop = dd(vAn, num);
        size_t sz = globalAns.size();
        
        if (dDop.first != 0) {
            Expr exprLst = vAn[dDop.first - 1].second;
            prove(vAn, globalAns, dDop.first - 1);

            for (const std::pair<bool, Expr>& it : dDop.second) {
                std::vector<tpair> tmp;
                if (!it.first) {
                    Impl imp = dynamic_cast<Impl>(exprLst);
                    exprLst = imp->right;
                    for (size_t j = sz; j < globalAns.size(); ++j) {
                        tmp.push_back(globalAns[j]);
                    }
                    globalAns.resize(sz);
                    globalAns.push_back({ exprLst, MOP });
                    globalAns.push_back({ it.second, HYPH });
                    for (const tpair& v : tmp) {
                        globalAns.push_back(v);
                    }
                }
                else {
                    exprLst = new implication(it.second, exprLst);
                    for (size_t i = globalAns.size(); i != 0 && (sz <= 0 || i != sz - 1);) {
                        i--;
                        Expr exprSEc = globalAns[i].first;

                        if (exprSEc->equals(it.second)) {
                            GetFirstType(it, tmp);
                        }
                        else if (globalAns[i].second == AXIOM || 
                            globalAns[i].second == HYPH) {
                            GetSecondType(it, tmp, exprSEc);
                        }
                        else {
                            GetThirdType(it, tmp, exprSEc, i, globalAns);
                        }
                    }
                    globalAns.resize(sz);
                    for (size_t i = 0; i < tmp.size(); i++) {
                        globalAns.push_back(tmp[tmp.size() - i - 1]);
                    }
                }
            }
        }
    }
}

std::string ans(const tvec& vAn) {
    std::stringstream result;
    std::vector<tpair> globalAns, gloablTmo;

    if (!vAn.empty()) {
        size_t count{ 0 };
        for (const Expr v : vAn.back().first) {
            if (count) result << ",";
            result << v->toString();
            count++;
        }

        result << "|-" << vAn.back().second->toString() << "\n";
        prove(vAn, globalAns, vAn.size() - 1);

        if (!globalAns.empty()) {
            rec_degradation(vAn, globalAns, gloablTmo, globalAns.front().first, 0);
            for (size_t i = 0; i < gloablTmo.size(); i++) {
                result << gloablTmo[gloablTmo.size() - i - 1].first->toString() << "\n";
            }
        }
    }
    return result.str();
}

int main() {
    std::string str;
    tvec vAn;
    while (std::getline(std::cin, str) && !str.empty()) {
        Parser parser;
        parser.init(str.c_str());
        std::vector<Expr> lft = parser.start_expr();
        Expr e = parser.expr();
        reinit(lft, e, false);
        vAn.push_back({ lft, e });
    }

    std::cout << ans(vAn);
    return 0;
}

File: term4/matlog/C/java/Main.java
import expression.*;
import expression.Variable;
import parser.Parser;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static final Integer AXIOM = 0x111;
    static final Integer MOP = 0x222;
    static final Integer HYPH = 0x333;

    public static class Pair<F, S> {
        public F first; //first member of pair
        public S second; //second member of pair

        public Pair() {
            this.first = null;
            this.second = null;
        }

        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }

    static Expression[] axioms = new Expression[10];
    private static Vector<Pair<Vector<Expression>, Expression>> van = new Vector<>();
    private static Vector<Pair<Expression, Integer>> globalAns = new Vector<>();
    private static Vector<Pair<Expression, Integer>> gloablTmo = new Vector<>();

    private static final Variable a = new Variable("a");
    private static final Variable b = new Variable("b");
    private static final Variable c = new Variable("c");

    private static void axInit() {
        axioms[0] = new Implication(a, new Implication(b, a));
        axioms[1] = new Implication(new Implication(a, b), new Implication(new Implication(a, new Implication(b, c)), new Implication(a, c)));
        axioms[2] = new Implication(a, new Implication(b, new Conjunction(a, b)));
        axioms[3] = new Implication(new Conjunction(a, b), a);
        axioms[4] = new Implication(new Conjunction(a, b), b);
        axioms[5] = new Implication(a, new Disjunction(a, b));
        axioms[6] = new Implication(b, new Disjunction(a, b));
        axioms[7] = new Implication(new Implication(a, c), new Implication(new Implication(b, c), new Implication(new Disjunction(a, b), c)));
        axioms[8] = new Implication(new Implication(a, b), new Implication(new Implication(a, new Negation(b)), new Negation(a)));
        axioms[9] = new Implication(new Negation(new Negation(a)), a);
    }

    private static Map<String, Expression> ntm = new HashMap<>();

    private static Expression reinit(Vector<Expression> lft, Expression e, boolean insert) {
        Expression tmp = e;
        while (tmp instanceof Implication) {
            Implication ptr = (Implication) tmp;
            if (insert) {
                lft.add(ptr.left);
            }
            tmp = ptr.right;
        }
        return tmp;
    }

    private static Expression reinit(Vector<Expression> lft, Expression e) {
        return reinit(lft, e, true);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String str;

        while ((str = in.readLine()) != null && str.length() > 2) {
            Parser parser = new Parser(str);
            van.add(new Pair<>(parser.start_expr(), parser.expr()));
        }

        axInit();
        answer();

    }

    private static int isAx(Expression expr) {
        for (int i = 0; i < 10; i++) {
            ntm.clear();
            if (axioms[i].eq(expr, ntm, false)) return i + 1;
        }
        return 0;
    }

    private static boolean cCheck(Vector<Expression> lft, Vector<Expression> rft) {
        Boolean[] dop = new Boolean[rft.size()];
        if (lft.size() <= rft.size()) {
            Arrays.fill(dop, false);
            for (Expression expression : lft) {
                int j = 0;
                for (; j <= rft.size(); j++) {
                    if (j < rft.size() && !dop[j] && j != rft.size() && expression.eq(rft.get(j), ntm)) {
                        dop[j] = true;
                        break;
                    }
                }
                if (j == rft.size() + 1) return false;
            }
            return true;
        }
        return false;
    }

    private static Pair<Integer, Integer> mCheck(int i) {
        for (int j = i; j >= 0; j--) {
            if (van.get(j).second instanceof Implication) {
                Implication dop = (Implication) van.get(j).second;
                if (dop.right.eq(van.get(i).second, ntm) && cCheck(van.get(j).first, van.get(i).first)) {
                    for (int k = i; k >= 0; k--) {
                        if (k != j && dop.left.eq(van.get(k).second, ntm)) {
                            if (cCheck(van.get(k).first, van.get(i).first)) {
                                return new Pair<>(j, k);
                            }
                        }
                    }
                }
            }
        }
        return new Pair<>(-1, -1);
    }

    private static void answer() {
        if (!van.isEmpty()) {
            for (int i = 0; i < van.get(van.size() - 1).first.size(); i++) {
                if (i != 0) System.out.print(",");
                System.out.print(van.get(van.size() - 1).first.get(i).toString());
            }

            System.out.print("|-" + van.get(van.size() - 1).second.toString() + "\n");


            prove(van.size() - 1);


            if (!globalAns.isEmpty()) {
                rec_degradation(globalAns.get(0).first, 0);
                for (int i = 0; i < gloablTmo.size(); i++) {
                    System.out.print(gloablTmo.get(gloablTmo.size() - i - 1).first.toString() + "\n");
                }
            }
        }
    }

    private static void rec_degradation(Expression e, int num) {
        for (int i = num + 1; i < globalAns.size(); i++) {
            if (globalAns.get(i).first.eq(e, ntm)) {
                rec_degradation(globalAns.get(i).first, i);
                return;
            }
        }

        for (int i = 0; i < van.lastElement().first.size(); i++) {
            if (e.eq(van.lastElement().first.get(i), ntm)) {
                gloablTmo.add(new Pair<>(van.lastElement().first.get(i), HYPH));
                return;
            }
        }
        if (isAx(e) == 0) {
            for (int i = num + 1; i < globalAns.size(); i++) {
                if (globalAns.get(i).first instanceof Implication) {
                    Implication dop = (Implication) globalAns.get(i).first;
                    if (dop.right.eq(e, ntm)) {
                        for (int j = num + 1; j < globalAns.size(); j++) {
                            if (i != j && dop.left.eq(globalAns.get(j).first, ntm)) {
                                gloablTmo.add(new Pair<>(e, MOP));
                                rec_degradation(globalAns.get(j).first, j);
                                rec_degradation(globalAns.get(i).first, i);
                                return;
                            }
                        }
                    }
                }
            }
        }
        gloablTmo.add(new Pair<>(e, AXIOM));
    }

    private static Pair<Integer, Vector<Pair<Boolean, Expression>>> dCheck(int num) {
        Vector<Expression> dop = van.get(num).first;
        Expression expr = reinit(dop, van.get(num).second);
        Vector<Expression> vec;
        Vector<Expression> kostil = new Vector<>();
        Vector<Pair<Boolean, Expression>> dop2 = new Vector<>();

        for (int i = num; i != 0; ) {
            i--;
            vec = van.get(i).first;

            if (expr.eq(reinit(vec, van.get(i).second), ntm) && cCheck(vec, dop)) {
                dop.clear();
                Expression tmp = van.get(i).second;
                for (; ; ) {
                    expr = van.get(num).second;
                    kostil.clear();
                    for (; ; ) {
                        if (expr.eq(tmp, ntm)) {
                            for (int j = 0; j < kostil.size(); j++) {
                                dop2.add(new Pair<>(true, kostil.get(kostil.size() - j - 1)));
                            }
                            return new Pair<>(i + 1, dop2);
                        }
                        if (expr instanceof Implication) {
                            Implication ptr = (Implication) expr;
                            expr = ptr.right;
                            kostil.add(ptr.left);
                        } else break;
                    }
                    if (tmp instanceof Implication) {
                        Implication ptr = (Implication) tmp;
                        tmp = ptr.right;
                        dop2.add(new Pair<>(false, ptr.left));
                    } else break;
                }
                break;
            }
        }

        return new Pair<>(0, new Vector<>());
    }

    private static void GetFirstType(Pair<Boolean, Expression> it, Vector<Pair<Expression, Integer>> tmp) {
        Implication t1 = new Implication(it.second, it.second);
        Implication t2 = new Implication(it.second, t1);
        Implication t3 = new Implication(it.second, new Implication(t1, it.second));
        Implication t4 = new Implication(t3, t1);

        {
            tmp.add(new Pair<>(t2, AXIOM));
            tmp.add(new Pair<>(new Implication(t2, t4), AXIOM));
            tmp.add(new Pair<>(t4, MOP));
            tmp.add(new Pair<>(t3, AXIOM));
            tmp.add(new Pair<>(t1, MOP));
        }
    }

    private static void GetSecondType(Pair<Boolean, Expression> it,
                                      Vector<Pair<Expression, Integer>> tmp, Expression expr) {
        Implication t1 = new Implication(it.second, expr);

        {
            tmp.add(new Pair<>(new Implication(expr, t1), AXIOM));
            tmp.add(new Pair<>(expr, HYPH));
            tmp.add(new Pair<>(t1, MOP));
        }
    }

    private static void GetThirdType(Pair<Boolean, Expression> it,
                                     Vector<Pair<Expression, Integer>> tmp, Expression expr,
                                     int num) {
        int cntr = 0, dop = 0;
        for (int i = num + 1; i < globalAns.size(); i++) {
            if (cntr > 0) break;
            if (globalAns.get(i).first instanceof Implication) {
                if (((Implication) globalAns.get(i).first).right.eq(globalAns.get(num).first, ntm)) {
                    for (int j = num + 1; j < globalAns.size(); j++) {
                        if (i != j && ((Implication) globalAns.get(i).first).left.eq(globalAns.get(j).first, ntm)) {
                            cntr = j + 1;
                            dop = i + 1;
                            break;
                        }
                    }
                }
            }
        }
        Implication t1 = new Implication(it.second, expr);
        Implication t2 = new Implication(new Implication(it.second, globalAns.get(dop - 1).first), t1);
        tmp.add(new Pair<>(new Implication(new Implication(it.second, globalAns.get(cntr - 1).first), t2), AXIOM));
        tmp.add(new Pair<>(t2, MOP));
        tmp.add(new Pair<>(t1, MOP));
    }

    private static void prove(int num) {
        Pair<Integer, Integer> mod = null;
        Pair<Integer, Vector<Pair<Boolean, Expression>>> dDop;
        int sz = 0;

        for (Expression expr : van.get(num).first) {
            if (expr.eq(van.get(num).second, ntm)) {
                globalAns.add(new Pair<>(van.get(num).second, HYPH));
                return;
            }
        }
        if (isAx(van.get(num).second) > 0) {
            globalAns.add(new Pair<>(van.get(num).second, AXIOM));
            return;
        }

            mod = mCheck(num);

        if (mod.first != -1 && mod.second != -1 && (mod.first != num && mod.second != num)) {
            globalAns.add(new Pair<>(van.get(num).second, MOP));
            prove(mod.first);
            prove(mod.second);
            return;
        }


        dDop = dCheck(num);
        sz = globalAns.size();

        if (dDop.first != 0 && dDop.first != num + 1) {
            prove(dDop.first - 1);
            Expression exprLst = van.get(dDop.first - 1).second;
            Vector<Pair<Expression, Integer>> tmp = new Vector<>();

            for (Pair<Boolean, Expression> it : dDop.second) {
                if (!it.first) {
                    exprLst = ((Implication) exprLst).right;
                    for (int j = sz; j < globalAns.size(); ++j) {
                        tmp.add(globalAns.get(j));
                    }
                    globalAns.setSize(sz);
                    globalAns.add(new Pair<>(exprLst, MOP));
                    globalAns.add(new Pair<>(it.second, HYPH));
                    globalAns.addAll(tmp);
                } else {
                    exprLst = new Implication(it.second, exprLst);

                    for (int i = globalAns.size(); i != 0 && (sz <= 0 || i != sz - 1); ) {
                        i--;

                        if (globalAns.get(i).first.eq(it.second, ntm)) {
                            GetFirstType(it, tmp);
                        } else if (globalAns.get(i).second == AXIOM ||
                                globalAns.get(i).second == HYPH) {
                            GetSecondType(it, tmp, globalAns.get(i).first);
                        } else {
                            GetThirdType(it, tmp, globalAns.get(i).first, i);
                        }
                    }
                    globalAns.setSize(sz);
                    for (int i = 0; i < tmp.size(); i++) {
                        globalAns.add(tmp.get(tmp.size() - i - 1));
                    }
                }
                tmp.clear();
            }
        }
    }
}
